zstd.cc

```diff
@@ -23,116 +23,106 @@ static const sstring COMPRESSOR_NAME = compressor::namespace_prefix + "ZstdCompr
 static const size_t DCTX_SIZE = ZSTD_estimateDCtxSize();
 
 class zstd_processor : public compressor {
-    int _compression_level = 3;
-    size_t _cctx_size;
-
-    static auto with_dctx(std::invocable<ZSTD_DCtx*> auto f) {
-        // The decompression context has a fixed size of ~128 KiB,
-        // so we don't bother ever resizing it the way we do with
-        // the compression context.
-        static thread_local std::unique_ptr<char[]> buf = std::invoke([&] {
-            auto ptr = std::unique_ptr<char[]>(new char[DCTX_SIZE]);
-            auto dctx = ZSTD_initStaticDCtx(ptr.get(), DCTX_SIZE);
-            if (!dctx) {
-                // Barring a bug, this should never happen.
-                throw std::runtime_error("Unable to initialize ZSTD decompression context");
-            }
-            return ptr;
-        });
-        return f(reinterpret_cast<ZSTD_DCtx*>(buf.get()));
-    }
+    int _compression_level;
 
-    static auto with_cctx(size_t cctx_size, std::invocable<ZSTD_CCtx*> auto f) {
-        // See the comments to reusable_buffer for a rationale of using it for compression.
-        static thread_local utils::reusable_buffer<lowres_clock> buf(std::chrono::seconds(600));
-        static thread_local size_t last_seen_reallocs = buf.reallocs();
-        auto guard = utils::reusable_buffer_guard(buf);
-        // Note that the compression context isn't initialized with a particular
-        // compression config, but only with a particular size. As long as
-        // it is big enough, we can reuse a context initialized by an
-        // unrelated instance of zstd_processor without reinitializing it.
-        //
-        // If the existing context isn't big enough, the reusable buffer will
-        // be resized by the next line, and the following `if` will notice that
-        // and reinitialize the context.
-        auto view = guard.get_temporary_buffer(cctx_size);
-        if (last_seen_reallocs != buf.reallocs()) {
-            // Either the buffer just grew because we requested a buffer bigger
-            // than its last capacity, or it was shrunk some time ago by a timer.
-            // Either way, the resize destroyed the contents of the buffer and
-            // we have to initialize the context anew.
-            auto cctx = ZSTD_initStaticCCtx(view.data(), buf.size());
-            if (!cctx) {
-                // Barring a bug, this should never happen.
-                throw std::runtime_error("Unable to initialize ZSTD compression context");
-            }
-            last_seen_reallocs = buf.reallocs();
-        }
-        return f(reinterpret_cast<ZSTD_CCtx*>(view.data()));
-    }
+    std::shared_ptr<ZSTD_CCtx> _cctx;
+    std::shared_ptr<ZSTD_DCtx> _dctx;
```

Comment:
Because our shared_ptr:s don't allow specific deleters. But we'd still like to share them between `replace`:ed instances.

Reply:
> Because our shared_ptr:s don't allow specific deleters.

`lw_shared_ptr` doesn't allow specifying deleters per object (while std::shared_ptr allows that), but it still allows specifying a deleter per type (by providing a specialization of  `lw_shared_ptr_deleter`), which would be enough in this case.

Not that the difference between `lw_shared_ptr` and `std::shared_ptr` is very important here, though.

Reply:
lw_shared_ptr cannot take ownership of an opaque pointer. It can only allocate a known object type.

Reply:
What about `seastar::shared_ptr`?

Reply:
Again, does not take deleters.

Reply:
> lw_shared_ptr cannot take ownership of an opaque pointer.

Ouch, right. I thought that zstd had a way to split the allocation and the initialization of `ZSTD_CCtx` (so we could allocate the shared pointer contents first, and initialize them later), but apparently it doesn't.